### PR TITLE
[Android] Improve SslStream PAL buffer resizing

### LINKED ISSUE #95295: Android SslStream splits messages over 2048 bytes in multiple TLS records (breaks MS-TDS protocol)
### Description

The Android SslStream implementation using the [Android SSLEngine](https://developer.android.com/reference/javax/net/ssl/SSLEngine#wrap(java.nio.ByteBuffer,%20java.nio.ByteBuffer)) APIs splits application data over 2048 bytes in multiple TLS records, something which *normally* isn't important, *except* for SQL Server which [expects TDS RPC requests (query messages) to be sent in a single TLS record](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-tds/619c43b6-9495-4a58-9e49-a4950db245b3), [breaking Microsoft.Data.SqlClient on Android](https://github.com/dotnet/SqlClient/issues/2141). This issue only affects SslStream Android implementation, all other platforms resize internal buffers to send larger TLS records that can be safely fragmented at the transport level instead.

### Reproduction Steps

Write a simple Android project that makes a TLS connection to a service and send a message of size larger than 2048 bytes - if it is encoded in multiple TLS *records*, the issue is successfully reproduced. As for reproducing the issue in a context where it matters enough to cause SQL Server to disconnect the client, a test project is available [in the original SqlClient issue](https://github.com/dotnet/SqlClient/issues/2141#issuecomment-1706271994). From there, all you need is to connect to SQL Server with TLS configured and enforced and make a query using a string longer than 956 characters (UTF-16 doubles the byte size + header overhead), resulting in a TDS RPC request longer than 2048 bytes over the wire. Upon receiving the message split in two TLS records, SQL Server drops the connection:

image

### Expected behavior

[SslStream.Write](https://learn.microsoft.com/en-us/dotnet/api/system.net.security.sslstream.writeasync?view=net-8.0) should send input buffers in the smallest number of output TLS records by default, resizing internal buffers accordingly, and taking into account the maximum TLS record data size (16384). This means that calling SslStream.Write with a buffer of 10000 bytes should result in a single TLS record even if internal buffer sizes were initialized to a smaller size. As long as we try sending the input buffers as complete TLS records instead of splitting them, edge cases like MS-TDS RPC requests sent to SQL Server should work.

### Actual behavior

SslStream.Write with input buffers larger than 2048 bytes are automatically split and sent in multiple TLS records, [breaking long SQL queries made with SqlClient on Android](https://github.com/dotnet/SqlClient/issues/2141).

### Regression?

This used to work in the old Xamarin.Android runtime prior to .NET Core (yes, the ancient one!), but since we've been waiting for .NET 6,7,8 for other [SslStream-related Android issues to be fixed](https://github.com/dotnet/runtime/issues/45741), we only noticed this particular regression after finally switching to .NET 8.

### Known Workarounds

Increasing the internal initial buffer size from 2048 to 4096 [is sufficient to work around the issue](https://github.com/dotnet/SqlClient/issues/2141#issuecomment-1827016651), and we have managed to inject a patched System.Net.Security.dll in our builds as a temporary solution. However, it's far from ideal and an upstream fix is needed in the .NET runtime, even if it means just increasing the initial buffer size upstream as a quick fix.

```
# https://github.com/dotnet/SqlClient/issues/2141
$SystemNetSecurityDllZipUrl = "https://github.com/dotnet/SqlClient/files/13468922/System.Net.Security.dll.zip"
Invoke-WebRequest -Uri $SystemNetSecurityDllZipUrl -OutFile "System.Net.Security.dll.zip"
Expand-Archive "System.Net.Security.dll.zip" -DestinationPath .
$AndroidRuntimePrefix = "C:\Program Files\dotnet\packs\Microsoft.NETCore.App.Runtime.Mono.android"
Get-Item "${AndroidRuntimePrefix}-*\8.0.0\runtimes\android-*\lib\net8.0\System.Net.Security.dll" | % {
  $BackupFile = ($_.FullName + ".bak")
  Move-Item $_ $BackupFile -Force
  Copy-Item "System.Net.Security.dll" $_ -Force
  @($_, $BackupFile) | % { Get-FileHash $_ } | Format-List
}
```

### Configuration

.NET 8 for Android - both x64 and ARM64. The same code using SqlClient on all other platforms (iOS, Windows, macOS, Linux) is unaffected by this issue.

### Other information

A lot of additional information is present [in the original SqlClient issue](https://github.com/dotnet/SqlClient/issues/2141).

### PR DESCRIPTION
**Pull request by simonrozsival**
Our current implementation of the `AndroidCryptoNative_SSLStreamWrite` doesn't correctly handle buffer overflows when calling the `wrap` method. This isn't an issue because we always create buffers large enough to fit the wrapped data. The problem is that we are splitting data into many small chunks instead of sending fewer larger TLS frames and reducing overhead. This also affects certain protocols such as MS-TDS as reported in #95295.

Closes #95295 

/cc @rzikm @wfurt

### CODE REVIEW

File: src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c [hunk 1]
```diff
@@ -858,26 +885,24 @@ PAL_SSLStreamStatus AndroidCryptoNative_SSLStreamWrite(SSLStream* sslStream, uin
...
     PAL_SSLStreamStatus ret = SSLStreamStatus_Error;
 
-    // int remaining = appOutBuffer.remaining();
-    // int arraySize = length > remaining ? remaining : length;
     // byte[] data = new byte[arraySize];
-    int32_t remaining = (*env)->CallIntMethod(env, sslStream->appOutBuffer, g_ByteBufferRemaining);
-    int32_t arraySize = length > remaining ? remaining : length;
-    jbyteArray data = make_java_byte_array(env, arraySize);
+    // data.setByteArrayRegion(0, length, buffer);
+    jbyteArray data = make_java_byte_array(env, length);
```
Comment (simonrozsival) [inline src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c:890]: I thought about capping the size of the `data` buffer to the max payload size of one TLS frame and keep copying just a subset of the data as we did previously, but the SslStream is already chunking the data (`StreamSizes.Default == 32,768`).

**Review by rzikm — APPROVED** [review]
LGTM

File: src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c (hunk 1 above)
Comment (vcsjones) [inline src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c:890]: Do we need to allocate a java array here or can we use NIO? `appOutBuffer` is a `ByteBuffer`. `ByteBuffer` has a `put` that accepts another `ByteBuffer`.

We can create a `ByteBuffer` over `buffer` with `jobject bufferByteBuffer = (*env)->NewDirectByteBuffer(env, buffer, length);`

Then we can `put` that in to appOutBuffer. That remove an allocation and a copy from a handshake.

Comment (vcsjones) [reply src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c:890]: Just a thought, can be done as a follow up.

Comment (simonrozsival) [reply src/native/libs/System.Security.Cryptography.Native.Android/pal_sslstream.c:890]: Thanks for the suggestion, I'll give it a try.

**Review by steveisok — APPROVED** [review]

**Comment by simonrozsival** [comment]
There are failing runtime-extra-platforms tests, but they are all unrelated:
- This PR only touches Android PAL code so all non-Android platforms cannot be affected
- On Android, the failing tests are known issues:
    - System.Net.Sockets.Tests: https://github.com/dotnet/runtime/issues/104709 and https://github.com/dotnet/dnceng/issues/3405 - connectivity issues
    - System.Net.Ping.Functional.Tests: unrelated test is failing (`UnixPingUtilityTests.TimeoutIsRespected`)
    - System.Net.NameResolution.Pal.Tests: unrelated test is faling (`NameResolutionPalTests.TryGetAddrInfo_ExternalHost`)

**Pull request merged**